src/events/rawGuildMemberUpdate.ts

```diff
@@ -22,34 +23,65 @@ export default class extends Event {
 		floatPromise(this, this.handleRoleSets(guild, data));
 	}
 
-	private handleNicknameChange(guild: KlasaGuild, data: WSGuildMemberUpdate) {
-		// Get the current nickname, compare them both, if they are different, it changed
+	private handleMemberChange(guild: KlasaGuild, data: WSGuildMemberUpdate) {
+
+		// Look up which logs should be send in the guild database
+		let shouldLogNickname = guild.settings.get(GuildSettings.Events.MemberNicknameUpdate);
+		let shouldSendRoleLog = guild.settings.get(GuildSettings.Events.MemberRoleUpdate);
+
+		// Get the currently stored dataset
 		const previous = guild.memberTags.get(data.user.id);
+
+		// Setup the next stored dataset
 		const next: MemberTag = {
 			nickname: data.nick || null,
 			joinedAt: typeof previous === 'undefined' ? null : previous.joinedAt,
 			roles: data.roles
 		};
 
-		// Get the previous nickname
+		// Store the next data set in the MemberTags store
 		guild.memberTags.set(data.user.id, next);
 
-		// If the previous was unset or it's the same as the next one, skip
-		if (typeof previous === 'undefined' || previous.nickname === next.nickname) return;
+		// If the previous was unset then skip all
+		if (typeof previous === 'undefined') return;
+
+		// If nicknames are identical then skip nickname log
+		if (previous.nickname === next.nickname) shouldLogNickname = false;
 
-		// TODO(kyranet): Role Change Logs
-		if (guild.settings.get(GuildSettings.Events.MemberNicknameUpdate)) {
+		if (shouldLogNickname) {
+			// Send the Nickname log
 			this.client.emit(Events.GuildMessageLog, MessageLogsEnum.Member, guild, () => new MessageEmbed()
 				.setColor(Colors.Yellow)
 				.setAuthor(`${data.user.username}#${data.user.discriminator} (${data.user.id})`, getDisplayAvatar(data.user.id, data.user))
 				.setDescription(guild.language.tget('EVENTS_NAME_DIFFERENCE', previous.nickname, next.nickname))
 				.setFooter(guild.language.tget('EVENTS_NICKNAME_UPDATE'))
 				.setTimestamp());
 		}
+
+		// If role arrays are identical then skip role log
+		if (arrayStrictEquals(previous.roles, next.roles)) shouldSendRoleLog = false;
+
+		if (shouldSendRoleLog) {
+			// Get the names of each role for logging
+			const previousRoleNames = previous.roles.length
+				? previous.roles.map(previousRoleId => `\`${guild.roles.get(previousRoleId)!.name}\``).join(', ')
+				: null;
+			const nextRoleNames = next.roles.length
+				? next.roles.map(nextRoleId => `\`${guild.roles.get(nextRoleId)!.name}\``).join(', ')
+				: null;
+
+			// Set the Role change log
+			this.client.emit(Events.GuildMessageLog, MessageLogsEnum.Member, guild, () => new MessageEmbed()
+				.setColor(Colors.Yellow)
+				.setAuthor(`${data.user.username}#${data.user.discriminator} (${data.user.id})`, getDisplayAvatar(data.user.id, data.user))
+				.setDescription(guild.language.tget('EVENTS_ROLE_DIFFERENCE', previousRoleNames, nextRoleNames))
+				.setFooter(guild.language.tget('EVENTS_ROLE_UPDATE'))
+				.setTimestamp());
+		}
```

Comment:
This should not log all roles, but the ones that were added or removed.
I think Skyra III had this code, might be worth a check? I can check in around 3 hours if I don't forget.

Reply:
Found it, v2.1.1:

https://github.com/skyra-project/skyra/blob/7d5a418bedfc026fefb2062774e284d0e79a423a/events/guildMemberUpdate.js#L59-L69

Code could be tidied up, it's a little messy, to be expected from +2 years old code.